src/Mod/Sketcher/Gui/TaskSketcherElements.h

```diff
@@ -39,21 +40,91 @@ namespace SketcherGui {
 class ViewProviderSketch;
 class Ui_TaskSketcherElements;
 
+enum class ElementType { edge, start, end, mid, none }; //This is to identify the type of element selected
+
+// helper class to store additional information about the listWidget entry.
+class ElementItem : public QListWidgetItem
+{
+public:
+
+    ElementItem(int elementnr, int startingVertex, int midVertex, int endVertex,
+        Base::Type geometryType, bool construction, bool isinternal, bool external, QIcon ic0, QIcon ic1, QIcon ic2, QIcon ic3, QString lab) :
+        ElementNbr(elementnr)
+        , StartingVertex(startingVertex)
+        , MidVertex(midVertex)
+        , EndVertex(endVertex)
+        , isLineSelected(false)
+        , isStartingPointSelected(false)
+        , isEndPointSelected(false)
+        , isMidPointSelected(false)
+        , GeometryType(geometryType)
+        , isConstruction(construction)
+        , isInternal(isinternal)
+        , isExternal(external)
+        , clickedOn(ElementType::none)
+        , hovered(ElementType::none)
+        , rightClicked(false)
+        , icon0(ic0)
+        , icon1(ic1)
+        , icon2(ic2)
+        , icon3(ic3)
+        , label(lab)
+    {}
+
+    ~ElementItem() override {
+    }
+
+    int ElementNbr;
+    int StartingVertex;
+    int MidVertex;
+    int EndVertex;
+    bool isLineSelected;
+    bool isStartingPointSelected;
+    bool isEndPointSelected;
+    bool isMidPointSelected;
+    Base::Type GeometryType;
+    bool isConstruction;
+    bool isInternal;
+    bool isExternal;
+
+    ElementType clickedOn;
+    ElementType hovered;
+    bool rightClicked;
+
+    QIcon icon0;
+    QIcon icon1;
+    QIcon icon2;
+    QIcon icon3;
+    QString label;
+};
+
+class ElementItemDelegate : public QStyledItemDelegate
+{
+    Q_OBJECT
+public:
+    explicit ElementItemDelegate(QObject* parent);
+    ~ElementItemDelegate() override;
+
+    void paint(QPainter* painter, const QStyleOptionViewItem& option, const QModelIndex& index) const override;
+    bool editorEvent(QEvent* event, QAbstractItemModel* model, const QStyleOptionViewItem& option, const QModelIndex& index) override;
+
+    ElementItem* getElementtItem(const QModelIndex& index) const;
+
+Q_SIGNALS:
+    void onMouseHover(QModelIndex);
+};
+
 class ElementView : public QListWidget
 {
     Q_OBJECT
 
 public:
     explicit ElementView(QWidget *parent = nullptr);
     ~ElementView() override;
-
-
-Q_SIGNALS:
-    void onFilterShortcutPressed();
+    ElementItem* itemFromIndex(const QModelIndex& index);

```

Comment:
Should be `override` if I got it well

Reply:
I wasn't sure because the function doesn't have the same return type than the QListWidget function. You think it should have?

Reply:
You're right ;)

Reply:
OK, according to https://stackoverflow.com/questions/8967303/override-a-member-function-with-different-return-type , this is a valid override (because the return type of the subclass function inherits the return type of the superclass function).

Reply:
Ok I'm adding it then.

Reply:
I can't add it. I get an error saying there's no function to override.

Reply:
The issue is that itemFromIndex is not a virtual function. Then the "override" marking does not apply here.

What happens is that the parent class function is hidden, but what is called when using a pointer depends from the actual type of the pointer (if it is a pointer the base class, or a pointer to the derived class).

I have a preference for not using the same name in this cases (it also has to do with the return type not being part of the signature of the function). So I would have renamed the function "elementItemFromIndex", which clearly conveys the idea you are getting the derived class version.

Reply:
Ok to rename it for me. Did you already did it ?

Reply:
> The issue is that itemFromIndex is not a virtual function. Then the "override" marking does not apply here.

Thanks for the analysis. Totally makes sense.
I agree function hiding should be avoided.